scalableminds/webknossos
# Add support for groups of trees

This PR adds support for tree groups to webknossos. Groups can be created/renamed/deleted/moved in the tree tab. Groups and trees can be moved using drag and drop. The existing functionality of the tree tab has been preserved (toggle checkbox to toggle visibility, select tree to set new active tree).
The groups are persisted and are also parsed-from/serialized-to NML files. Please see the `Mailable description` for the NML format.

Implementation-wise, the tree groups are stored as a separate hierarchical data structure and trees have a new `groupId` property that is either `null` if the tree is not part of any group or contains the groupId. I implemented it this way to keep the overall code changes small.

I've added unit tests for the frontend nml group serializing/parsing functionality.

Missing (@fm3 please assign this issue to someone to implement the missing backend functionality):
  - [x] Backend should serialize a tree's groupId and the groups itself to NMLs (see `Mailable description` for the NML format)
  - [x] Backend should parse a tree's optional groupId as well as (potentially existing) groups from NMLs
  - [x] Backend should update a tree's groupId if it is updated as part of an `updateTree` update action - groupId is either a string or null
  - [x] Backend should update the treeGroups that are sent as part of a new `updateTreeGroups` update action. As of now, the whole treeGroups object is sent every time (we can always add more specific update actions later). It looks like this:
```javascript
type TreeGroupType = {
  name: string,
  groupId: string,
  children: Array<TreeGroupType>,
};

type UpdateTreeGroupsUpdateAction = {
  name: "updateTreeGroups",
  value: {
    treeGroups: Array<TreeGroupType>,
  },
};
```

Please let me know if you're not happy with the data format or need clarifications :)

### Mailable description of changes:
 - Groups of trees (tree hierarchies) are now supported by webknossos. Use drag and drop to organize trees into groups in the tree tab. Hover over a group to access the group management menu to create/rename/delete groups. Groups are also parsed-from/serialized-to NMLs like this to maintain backwards-compatibility:
```xml
<things>
    <thing groupId="1" name="Axon 1">...</thing>
    <thing groupId="1" name="Synapse 1a">...</thing>
    <thing groupId="1" name="Synapse 1b">...</thing>
    <thing groupId="2" name="Axon 2">...</thing>
    <thing groupId="2" name="Synapse 2a">...</thing>
<things>
<groups>
  <group id="1" name="Axon 1">
      <group id="3" name="Blah" />
  </group>
  <group id="2" name="Axon 2" />
</groups>
```

### Steps to test:
- Run `yarn test`
- Open a tracing, create groups in the tree tab, move trees into groups, move groups into groups, delete groups, rename groups
- Save the tracing and reload the page,  you should see the same tracing with the same trees in the same groups
- Download a tracing with groups using the Frontend and Backend NML download
- Upload it again, you should see the same tracing with the same trees in the same groups
- Upload an old NML without any groups - this should still work fine

### Issues:
- fixes #2514

------
